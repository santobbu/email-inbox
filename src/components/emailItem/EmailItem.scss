.EmailItem {
    padding: 12px 6px;

    &:hover {
        cursor: pointer;
        background-color: #F8F8F8;
    }

    &--selected {
        background-color: #F8F8F8;
    }

    .ficon {
        margin: 0 4px;
    }

    &__Content {
        line-height: 16px;
    }

    &__Subject {
        display: inline-block;
        width: calc(100% - 100px);
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
        max-width: 790px;
        font-weight: 500;
    }

    &__Time {
        display: inline-block;
        width: 100px;
        text-align: right;
    }

    &__Body {
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
        max-width: 900px;
    }

    &__Name {
        display: inline-block;
        width: calc(100% - 200px);
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
        max-width: 690px;
        font-weight: 700;
        margin-bottom: 4px;
    }

    &__Action {
        display: inline-block;
        width: 200px;
        text-align: right;
        vertical-align: top;
    }

    .ant-avatar {
        font-size: 12px;
        
        &--blue {
            background-color: blue;
        }

        &--pink {
            background-color: pink;
        }

        &--green {
            background-color: green;
        }

        &--brown {
            background-color: #A52A2A;
        }

        &--skyBlue {
            background-color: #00BFFF;
        }
    }
}