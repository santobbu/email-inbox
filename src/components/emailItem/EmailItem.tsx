
import * as React from 'react'
import { List, Avatar  } from 'antd';
import { EmailItem as EmailItemTypes } from '../../types/Emails';
import './EmailItem.scss';

interface Props extends EmailItemTypes {
    isSelected: boolean;
    index: number;
    onItemClick: (event: React.MouseEvent<HTMLInputElement>, index: number) => void;
    color: string;
}

const EmailItem: React.FC<Props> = React.memo(props => {

    const onDeleteClicked = () => {};
    const onOpenClicked = () => {};
    const onMarkClicked = () => {};
    const onPinClicked = () => {};

    const renderEmailAction = () => {
        if (!props.isSelected) return null;

        return (
            <div className="EmailItem__Action">
                <i className="ficon icofont-bin" onClick={onDeleteClicked} />
                <i className="ficon icofont-envelope" onClick={onOpenClicked} />
                <i className="ficon icofont-flag" onClick={onMarkClicked} />
                <i className="ficon icofont-tack-pin" onClick={onPinClicked} />
            </div>
        );
    ;}
    
    const renderItemContent = () => {
        return (
            <div className="EmailItem__Content">
                <div className="EmailItem__Name">{props.from.name}</div>
                {renderEmailAction()}
                <div className="EmailItem__Subject">{props.subject}</div>
                <div className="EmailItem__Time">{props.time}</div>
                <div className="EmailItem__Body">{props.body}</div>
            </div>
        );
    };

    const onItemClick = (event: React.MouseEvent<HTMLInputElement>) => {
        props.onItemClick(event, props.index);
    }

    const getAvatar = () => {
        if (props.isSelected) {
            return <i className="icofont-check-circled"></i>;
        }

        var matches = props.from.name.match(/\b(\w)/g);
        var acronym = matches && matches.join('').slice(0, 2);

        return (
            <div>{acronym}</div>
        );
    }

    const avatarClass = `ant-avatar--${props.color}`;
    const selectedAvatar = props.isSelected ? 'ant-avatar--selected' : '';

    return (
        <List.Item className={`EmailItem ${props.isSelected ? 'EmailItem--selected' : ''}`} onClick={onItemClick}>
            <List.Item.Meta
                avatar={
                    <Avatar children={getAvatar()} className={`${avatarClass} ${selectedAvatar}`}/>
                }
                description={renderItemContent()}
            />
        </List.Item>
    );
})

EmailItem.displayName = 'EmailItem'
export default EmailItem