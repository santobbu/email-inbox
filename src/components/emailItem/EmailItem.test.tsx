import React from 'react';
import { shallow, configure  } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });


import EmailItem from './EmailItem';

describe('Email Item', () => {
    const props = {
        from: {
            name: 'Tester',
            email: 'Tester@gmail.com',
        }, 
        subject: 'This is test subject',
        body: 'This is email body',
        time: '10:00',
        isSelected: false,
        index: 0,
        onItemClick: () => {},
        color: 'blue',
    };

    it('should render component correctly for normal state', () => {
        const wrapper = shallow(<EmailItem {...props}/>);

        expect(wrapper).toMatchSnapshot();
    });

    it('should render component correctly for selected state', () => {
        const newProps = { ...props, isSelected: true }
        const wrapper = shallow(<EmailItem {...newProps}/>);

        expect(wrapper).toMatchSnapshot();
    });
});
