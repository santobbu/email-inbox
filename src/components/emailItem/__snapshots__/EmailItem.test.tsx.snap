// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email Item should render component correctly 1`] = `ShallowWrapper {}`;

exports[`Email Item should render component correctly for normal state 1`] = `ShallowWrapper {}`;

exports[`Email Item should render component correctly for selected state 1`] = `ShallowWrapper {}`;
